piff/gsobject_model.py

```diff
@@ -0,0 +1,389 @@
+
+        return dflux, dcentroid.x, dcentroid.y, dsize, dshape.g1, dshape.g2, flag
+
+    def getProfile(self, params):
+        """Get a version of the model as a GalSim GSObject
+
+        :param params:      A numpy array with either  [ size, g1, g2 ]
+                            or  [ cenu, cenv, size, g1, g2 ]
+                            depending on if the center of the model is being forced to (0.0, 0.0)
+                            or not.
+
+        :returns: a galsim.GSObject instance
+        """
+        if self._force_model_center:
+            size, g1, g2 = params
+            cenu, cenv = (0.0, 0.0)
+        else:
+            cenu, cenv, size, g1, g2 = params
+        return self.gsobj.dilate(size).shear(g1=g1, g2=g2).shift(cenu, cenv)
+
+    def draw(self, star):
+        """Draw the model on the given image.
+
+        :param star:    A Star instance with the fitted parameters to use for drawing and a
+                        data field that acts as a template image for the drawn model.
+
+        :returns: a new Star instance with the data field having an image of the drawn model.
+        """
+        prof = self.getProfile(star.fit.params).shift(star.fit.center) * star.fit.flux
+        image = star.image.copy()
+        prof.drawImage(image, method=self._method, offset=(star.image_pos-image.trueCenter()))
+        data = StarData(image, star.image_pos, star.weight, star.data.pointing)
+        return Star(data, star.fit)
+
+    def _lmfit_resid(self, lmparams, star):
+        """Residual function to use with lmfit.  Essentially `chi` from `chisq`, but not summed
+        over pixels yet.
+
+        :param lmparams:  An lmfit.Parameters() instance.  The model.
+        :param star:    A Star instance.  The data.
+
+        :returns: `chi` as a flattened numpy array.
+        """
+        image, weight, image_pos = star.data.getImage()
+        flux, cenu, cenv, size, g1, g2 = lmparams.valuesdict().values()
+        # Fit cenu and cenv regardless of force_model_center.  The difference is whether the fit
+        # value is recorded (force_model_center=False) or discarded (force_model_center=True).
+        prof = self.gsobj.dilate(size).shear(g1=g1, g2=g2).shift(cenu, cenv) * flux
+        model_image = image.copy()
+        prof.drawImage(model_image, method=self._method,
+                       offset=(image_pos - model_image.trueCenter()))
+        return (np.sqrt(weight.array)*(model_image.array - image.array)).ravel()
+
+    def _lmfit_params(self, star, vary_params=True, vary_flux=True, vary_center=True):
+        """Generate an lmfit.Parameters() instance from arguments.
+
+        :param star:         A Star from which to initialize parameter values.
+        :param vary_params:  Allow non-flux and non-center params to vary?
+        :param vary_flux:    Allow flux to vary?
+        :param vary_center:  Allow center to vary?
+
+        :returns: lmfit.Parameters() instance.
+        """
+        import lmfit
```

Comment:
Ah, yes.  Probably so.

We could consider making dependencies optional at some point if our dependency burden becomes too significant, but for now let's try to make sure we explicitly include everything that is needed for the unit tests to run properly.